File: Java/hello_world.java
import packageName;
import jave.awt.event.*;    // import all classes of Abstract Widget Toolkit (awt)
import javax.swing.JFrame;  // import only one class
// importing a package doesn't import its sub-packages

public class hello_world {
    static final void main (String [] args) throws Exception {
	System.out.println("Hello World!");
	System.out.println("command line argument one: " + args[0]);	// the first argument, not the command name

	System.out.println("Please input a number: ");
	Scanner scanner = new Scanner(System.in);
	int num = scanner.nextInt();
    }

    static void time () {
	long time = System.currentTimeMillis();
	Runtime runtime = Runtime.getRuntime();
    }
}

class primitive_data_type {
    // boolean
    boolean bool = true || false;   // 1 byte

    // Numeric
    /// char
    char x;	// 2 bytes -- support for unicode
	x = 'v';

    /// Int
    byte vb;	// 1 byte
    short vs;	// 2 bytes
    int vi;	// 4 bytes  
	vi = 0; vi = 10_000;   
    long vl;	// 8 bytes

    /// float
    float vf;	// 4 bytes	
	vf = 3.14f;	
    double vd;	// 8 bytes	
	vd = 20.22d;	

    // data casting
    ///* smaller capacity is assigned to a var of bigger capacity -- implicit
    vd = vi;
    ///* larger capacity to smaller capacity -- explicit
    vi = (int) vd;
}

class string {
    char [] arr = {'H', 'e', 'l', 'l', 'o'};
    String str1 = new String(arr);

    String str2 = "Hello World!";   // The JVM creates a memory location especially for Strings called String Constant Pool. That’s why String can be initialized without ‘new’ keyword.

    // methods
    str2.charAt(0); // 0 based
    str2.indexOf('H', int startIndex = 0);
    str2.indexOf("He");
    str2.lastIndexOf('H');
    str2.lastIndexOf("He");

    str2.length();
    str2.contains("substr");
    str2.startsWith("perfix");
    str2.endsWith("suffix");
    str2.replace("substr", "substitution"); // replace all occurrence
    str2.replaceAll("regexp", "substitution");
    str2.replaceFisrt("regexp", "substitution");
    str2.toLowerCase();
    str2.toUpperCase();

    String str3 = str1.concat(str2);
    String str4 = str1 + str2;

    /// comparison
    str2.compareTo("string");
    str2.compareToIgnoreCase("string");

    // splitting
    String str = "alpha, beta, delta, gamma, sigma";
    String arrSplit = str.split("regexp", int limit = 0);	// if the limit is omitted or zero, it will return all the strings splitted by the regex; otherwise, returns only required number of components without splitting the last component
    arrSplit = str.split("\\s");    // split by space
    
    // string to int
    String val = "100";
    int num1 = Integer.parseInt(val);
    int num2 = Integer.valueOf(val);
    // invalid number string will throw a NumberFormatException Error
}


class array {
    // declaration
    type[] arrayName;
    type arrayName[];

    type arrayName = new type[]; // declaration and construction

    type arrayName[] = {val1, valu2, valu3 ...};    // declaration and initialization

    // multi-dim arrays
    int[][] twoD = new int[4][4];

    // methods
    array.length;
    
}

class ArrayList { // size variable array
    ArrayList<type> a = new ArrayList<type>();

    // methods
    type o;
    a.add(o);
    a.remove(o);    // remove an element
    a.remove(0);    // remove element by index
    a.size();
    a.contains(o);  // returns boolean

}

class HashMaps {
    // it stores only object reference. so we can't use primitive data type like double or int; use wrapper class like Integer or Double instead
    HashMap<class1, class2> map = new HashMap<class1, class2>();
    map.get(key);
    map.put(key, class2 value);
    map.remove(key);	// remove an element 
    map.containsKey(key);
    map.isEmpty();
    map.keySet();   // all available keys
    map.values();   // all available values
}
class control_flow {
    // for
    for(i=0; i<5; i++) {
	System.out.println(i);
    }

    String arr[] = {"Hello", "World"};
    for (String ele : arr) {
	...
    }

    while (i<10) {
	...
    }

    do {
	...
    } while (i<10);

    // switch case
    int Switch = 4;
    swithc(Switch) {	// swtich can take input only as int or char, string is supported in new Java version 
	case 0:
	    ...;
	    break;
	case 1:
	    ...;
	    break;
	default:
	    ...;
	    break;
    }
}

class Class {
    //* static variable belongs to the class, rather than an object; they will be initialized before the initialization of any instance. Then a simply copy will be shared by all instances of this class, it can be accessed directly by the class name and don't need any object.
    //* static method belongs to the class, rather than an object; a static method can access only static data, it can not access non-static data (instance variables). A static method can call only other static methods and can not call a non-static method from it. A static method can be accessed directly by the class name and doesn't need any object; a static method can't refer to "this" or "super" keywords in anyway
    //* static block is a block of statement inside a Java class that will be executed when a class is first loaded into the JVM


    // Instance variable is declared inside a class but not inside a method
    int a;

    void set(int a) {
	// Local variable are declared inside a method including method arguments.
	this.a = a;
    }

    // constructor: 
    //* it has the same name as the class
    //* it should not return a value not even void

    // If no user-defined constructor is provided for a class, compiler initializes member variables to its default values.
    //* numeric data types are set to 0
    //* char data types are set to null character('\0')
    //* reference variables are set to null
    Class() {
	...
    }
}

class Class1 extends Class {	// class inheritance
    //* The "super" keyword: it is similar to "this" keyword, which can be used to access any data member or methods of the parent class

    //* the overridden method can widen the accessibility (private -> public) but not the vice versa. 

    //* dynamic polymorphism: If a base class reference (it may refer to an object of a child class) is used to call a method, the method to be invoked is decided by the JVM during run-time, depending on the object the reference is pointing to

    //* how to refer to parents methods when the parents and grandparents have the same name methods? still using super? how to distinguish parents and grandparents?

    // constructor chaining
    Class1() {
	super(arguments);   // the super statement needs to be the first line of code in the constructor of the child class
    }
}

abstract class Abstract {
    // An abstract class can't be instantiated
    // a class must be compulsorily labeled abstract, if it has one or more abstract methods

    // abstract method: method with only declaration
    abstract public void method();
}

interface Interface{
    // An interface is just like Java Class, but it only has static constants and abstract method. Java uses Interface to implement multiple inheritance. All methods in an interface are implicitly public and abstract.

    // Java doesn't allow multiple inheritance (can't extend more than 1 class), but an interface can extend from one or many interfaces. So the walkaround is interface. A class can implement multiple interfaces. It is necessary that the class must implement all the methods declared in the interfaces. The class cannot implement two interfaces in java that have methods with same name but different return type.

    // in interface, only one specifier is used -- public
    // the interface can't contain data fields
    // An interface cannot declare constructors or destructors.

    // An interface which is declared inside another interface is referred as nested interface
    
    // At the time of declaration, interface variable must be initialized. Otherwise, the compiler will throw an error.
}

class imp implements Interface {    // to use an interface in a class
}

class Error_n_Exception {
    /* There are two types of errors:
     * 1. compile time errors -- syntax errors, semantic errors
     * 2. runtiem errors: exception
     */
    try {
	statements;
    } catch (exceptionType) {
	...
    } catch (anotherException) {
	...
    } finally {
	...
    }

    // It may cumbersome to catch every excetpion, so Java use throw exception syntax:
    public static void method() throws IOException ArrayIndexBoundExcetpion {	// note here: we use throws (not throw) in method definition
	// By declaring the method may throw an excetpion
    }
}

class myException extends Exception {	// user-defined exception
    myException(int b) {
	...
    }

    public String toString() {
	return ("myException!");
    }


    try {
	throw new myException(2);
    } catch (myException e){
	System.out.println(e)
    }
}

File: Go/hello_world.go
package main

import "fmt"
import ("time"
	"math/rand"
	"math/cmplx"
	"io"
	"strings"
)

// when not explicitly initialized, var are given zero value of their type: 
// false for boolean, 0 for numeric and "" for string, nil for pointer
var t, f bool = true, false

var x int, x8 int8, x16 int16, x32 int32, x64 int64
var y uint, y8 uint8, y16 uint16, y32 uint32, y64 uint64
var z byte  // alias for uint8
var u rune  // alias for int32; Unicode code point
var v1 float32, v2 float64
var w1 complex64, w2 complex128 = cmplx.Sqrt(-5 + 12i)

var s string = "hello world"

// type conversion is explicit: Type(value)
v1 = float32(u)

// constants
const Pi = 3.14	// constant can't be declared using the := syntax
const (
    Big = 1 << 100
    Small = Big >> 99
)

// pointer
var i int = 42
var p *int = &i
*p = 21
fmt.Println(*p)


// struct: collection of fields
type Vertex struct {
    X int
    Y int
}
var v Vertex = Vertex{1, 2}
v.X = 4
/// pointer of vertex
var p *Vertex = &v
(*p).Y = 10 // this is cumbersome
p.Y = 10    // syntax sugar: go will interpret it as (*p).Y = 10

// array and slice
func array() {
    // arrays: fixed size
    var arr [5]int = [1, 2, 3, 4, 5]
    arr[0] = 6	// 0 based
    /// array literal
    p := [6]int{2, 3, 4, 5, 6, 7}


    // slice: varied size
    var slice []int = arr[1:4]	// half open: include the first one, but exclude the last one
    /// slice doesn't store any data, it likes a view of the underlying array; change the elements of a slice modifies the corresponding elements of the underlying array; other slices that share the same array will also see the changes
    /// make: create a zeroed array and return a slice that refers to that array
    a := make([]int, 5)	// len(a)=5
    b := make([]int, 0, 5)  // len(b)=0, cap(b)=5

    /// slice literal
    var s []int	// nil slice
    s == nil
    q := []int{2, 3, 5, 7, 11, 13}
    r := []bool{true, false, true, false, true, false}
    s := []struct {
	i int
	b bool
    }{
	{2, true},
	{3, false},
	{5, true},
	{7, true},
	{11, false},
	{13, true},
    }

    /// default bounds
    var a [10]int
    a[0:10] == a[:10] == a[0:] == a[:]

    /// length and capacity of a slice
    len(s)  // the number of elements a slice contains
    cap(s)  // the number of elements in the underlying array, counting from the first element in the slice

    /// re-slicing
    s := []int{2, 3, 5, 7, 11, 13}  // len=6; cap=6
    s = s[:0]	// len=0; cap=6
    s = s[:4]	// len=4; cap=6; here we can extend the slice because the underlying capicity is 6
    s = s[2:]	// len=2; cap=4

    /// slices of slices
    board := [][]string {
	[]string{"_", "_", "_"},
	[]string{"_", "_", "_"},
	[]string{"_", "_", "_"},
    }

    /// appending to a slice
    append(s []T, vs ... T) []T	    // if the backing array of s is too small to fit all the given values, a bigger array will be allocated, and then return slice that refers to the newly allocated array

    /// iterate over a slice (map): range
    pow := []int{1, 2, 4, 8, 16, 32, 64, 128}
    for i, v := range pow {
	...
    }
    for i := range pow {    // index only, omit the value
	...
    }
}

func maps() {
    // a nil map has no keys, nor can keys be added
    var m map[string]int
    m = make(map[string]int)
    m["Bobs"] = 100

    // map literals
    var m = map[string]int{
	"Bell": 10,
	"Google": 20,
    }

    // delete an element
    delete(m, key)

    elem, ok = m[key]
    // if key is in m, ok is true; otherwise ok is false, elem is the zero value for the map's element type
}

func main() {
    w := "word"	// short assignment, only available inside a function, declare with implicit type
    fmt.Println("Hello, World!")
    fmt.Println("The time is", time.Now())
    fmt.Println(rand.Intn(10))
    math.Sqrt(7)
    a, b = swap("hello", "world")
}

Exported    // only Capitalized names are exported (can be used by other packages)

// functions: arguments passed as value
/// post-parameter type
func add(x int, y int) int {
    return x + y
}
/// consecutive parameter type
func add(x, y int) int {    // omit the consecutive parameter type
    return x + y
}
/// function literal: kind of anonymous function
func(a, b int) { return a*b}(3, 4)

/// multiple returns
func swap(x, y string) (string, string) {
    return y, x
}

/// named return 
func split(sum int) (x, y int) {
    x = sum * 4 / 9
    y = sum - x
    return  // a return statement without arguments returns the named return values!!! BAD, don't do this
}

/// Defer: defer the execution of a function until the surrounding func returns; but the arguments are evaluated immediately
/*  1. A deferred function's arguments are evaluated when the defer statement is evaluated
    2. defer stack: LIFO
    3. deferred functions mya read and assign to the returning value
*/
func defer_func () {
    defer fmt.Println("world")

    fmt.Println("Hello")

    // defer stack: all deferred funcitons all push into a stack
    fmt.Println("counting")

    for i:= 0; i<10; i++ {
	defer fmt.Println(i)
    }

    fmt.Println("done")
}

fun c() (i int) {
    defer func() { i++ }()  // the return value is 2
    return 1
}
/// func as function argument
func compute(fn func(float64, float64) float64) float64 {
    return fn(3, 4)
}

/// closures
/// a closure is a function that references variables from outside its body. The function may access and assign to the referenced variables
func adder() func(int) int {	// return a closure
    sum := 0
    return func(x int) int {
	sum += x
	return sum  // each closure is bound to its own sum variable
    }
}

// control flows
/// if
if x < 0 {  // no parentheses, but {} is required
    ...
} else {
    ...
}
//// initialization for if
if v :=  math.Pow(x, n); v < lim {
    ...
}


/// switch
switch os := runtime.GOOS; os {
case "darwin":
    ...
case "linux":
    ...
default:
    ...
}
//// switch with no condition: which equals to switch true
switch {    // switch true
case 1:
    ...
case 2:
    ...
default:
    ...
}
/// loop: for
for i := 0; i<10; i++ {	// no parentheses around conditions; {} are required
    sum += i
}

/// simulated while
for sum < 1000 {
    sum += sum
}

//// forever running
for {
    ...
}

func methods() {
    // methods: go doesn't have classes. However, you can define methods on types
    /// A method is a function with a special receiver argument (like the self/this in other programming language), which appears between the func keyword and the method name
    func (v Vertex) Abs() float64 {
	return math.Sqrt(v.X*v.X + v.Y*v.Y)
    }
    /// you can only decalre a method with a receiver whose type is defined in the same package as the method; you can't declare a method with a receiver whose type is defined in another package (which includes the built-in types such as int)
    /// so if you want to declare a method for some built-in type; you need to create your own version
    type MyFloat float64
    func (f MyFloat) Abs() float64 {
	if f < 0 {
	    return float64(-f)
	}
	return float64(f)
    }

    /// pointer receivers
    /// the pointer must be the pointer of some *real* type, rather than pointer of a pointer
    /// methods with pointer receivers can modify the value to which the receiver points. while value receiver can't
    func (v *Vertex) Scale(f float64) {
	v.X = v.X * f
	v.Y = v.Y * f
    }

    /// calling a pointer receiver
    v := Vertex{3, 4}
    p := &Vertex{4, 3}
    v.Scale(3)	// syntax sugar: go will interpret it as (&v).Scale(3)
    p.Scale(3)
    /// the reverse also works: call a value receiver method with a pointer
    v.Abs()
    p.Abs() // go will interpret it as (*p).Abs()

    /// interface: an interface type is defined as a set of method signatures; a value of interface type can hold any value that implements those methods
    type Abser interface {
	Abs() float64
    }

    var a Abser
    f :=  float64(-math.Sqrt(2))
    v := Vertex{3, 4}

    a = f   // a float64 implementation of Abser
    a = &v  // a *Vertex implementation of Abser

    a.Abs()

    func (f float64) Abs() float64 {
	if f<0 {
	    return float64(-f)
	}
	return float64(f)
    }

    func (v *Vertex) Abs() float64 {
	return math.Sqrt(v.X*v.X + v.Y*v.Y)
    }

    /// an interface value can be though of as a tuple of (value, type) with the value in tuple being the concrete value of specific type
    /// nil value: an interface value that hold nil concrete value of some type is itself-non nil
    /// a nil interface value holds neither value nor concrete type. Calling a method on a nil interface is a run-time error 

    /// empty interface
    var i interface{}
    describe(i)	// (<nil>, <nil>)

    i = 42
    describe(i)	// (42, int)
    /// An empty intreface may hold values of any type (every type implements at least zero methods)

    /// type assertion: return an interface value's underlying concrete value
    t := i.(T)	    // if i doesn't hold type T, it will trigger a panic
    t, ok := i.(T)  // test if i holds type T; if not, t will be the zero value of type T and ok being false; and no panic occurs

    /// type switch: test type in series
    switch v := i.(type) {
    case T:
	...
    case S:
	...
    default:
	...
    }
}
